# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/printing/nsPrintJob.cpp
# Commit: 0347d8673dee
# Full Hash: 0347d8673deea522ae15120fe89e2552bf822a6d
# Author: Emily McDonough <emcdonough@mozilla.com>
# Date: 2022-12-03 09:24:59
# Regressor Bug: 1800546
# File Overlap Count: 1
# Description:
#   Bug 1800546 Part 1 - Use the style given the first page name for setting default orientation when printing r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D162058
# ==============================================================================

diff -r 0cda34ba7391 -r 0347d8673dee layout/printing/nsPrintJob.cpp
--- a/layout/printing/nsPrintJob.cpp	Sat Dec 03 01:15:55 2022 +0200
+++ b/layout/printing/nsPrintJob.cpp	Fri Dec 02 23:22:50 2022 +0000
@@ -1328,26 +1328,8 @@
   if (mPrintSettings->HasOrthogonalSheetsAndPages()) {
     std::swap(pageSize.width, pageSize.height);
   }
-
-  // If the document has a specified CSS page-size, we rotate the page to
-  // reflect this. Changing the orientation is reflected by the result of
-  // FinishPrintPreview, so that the frontend can reflect this.
-  // The new document has not yet been reflowed, so we have to query the
-  // original document for any CSS page-size.
-  if (const Maybe<StylePageSizeOrientation> maybeOrientation =
-          aPO->mDocument->GetPresShell()
-              ->StyleSet()
-              ->GetDefaultPageSizeOrientation()) {
-    if (maybeOrientation.value() == StylePageSizeOrientation::Landscape &&
-        pageSize.width < pageSize.height) {
-      // Paper is in portrait, CSS page size is landscape.
-      std::swap(pageSize.width, pageSize.height);
-    } else if (maybeOrientation.value() == StylePageSizeOrientation::Portrait &&
-               pageSize.width > pageSize.height) {
-      // Paper is in landscape, CSS page size is portrait.
-      std::swap(pageSize.width, pageSize.height);
-    }
-  }
+  // XXXalaskanemily: Is this actually necessary? We set it again before the
+  // first reflow.
   aPO->mPresContext->SetPageSize(pageSize);
 
   int32_t p2a = aPO->mPresContext->DeviceContext()->AppUnitsPerDevPixel();
@@ -1375,8 +1357,46 @@
   MOZ_TRY(aPO->mPresShell->Initialize());
   NS_ASSERTION(aPO->mPresShell, "Presshell should still be here");
 
+  RefPtr<PresShell> presShell = aPO->mPresShell;
+  {
+    // Get the initial page name. Even though we haven't done any page-name
+    // fragmentation (that happens during block reflow), this will still be
+    // valid to find the first page's name.
+    const nsAtom* firstPageName = nsGkAtoms::_empty;
+    if (const Element* const rootElement = aPO->mDocument->GetRootElement()) {
+      if (const nsIFrame* const rootFrame = rootElement->GetPrimaryFrame()) {
+        firstPageName = rootFrame->ComputePageValue();
+      }
+    }
+
+    // If the document has a specified CSS page-size, we rotate the page to
+    // reflect this. Changing the orientation is reflected by the result of
+    // FinishPrintPreview, so that the frontend can reflect this.
+    // The new document has not yet been reflowed, so we have to query the
+    // original document for any CSS page-size.
+    if (const Maybe<StylePageSizeOrientation> maybeOrientation =
+            presShell->StyleSet()->GetDefaultPageSizeOrientation(
+                firstPageName)) {
+      switch (maybeOrientation.value()) {
+        case StylePageSizeOrientation::Landscape:
+          if (pageSize.width < pageSize.height) {
+            // Paper is in portrait, CSS page size is landscape.
+            std::swap(pageSize.width, pageSize.height);
+          }
+          break;
+        case StylePageSizeOrientation::Portrait:
+          if (pageSize.width > pageSize.height) {
+            // Paper is in landscape, CSS page size is portrait.
+            std::swap(pageSize.width, pageSize.height);
+          }
+          break;
+      }
+      mMaybeCSSPageLandscape.emplace(maybeOrientation.value() ==
+                                     StylePageSizeOrientation::Landscape);
+      aPO->mPresContext->SetPageSize(pageSize);
+    }
+  }
   // Process the reflow event Initialize posted
-  RefPtr<PresShell> presShell = aPO->mPresShell;
   presShell->FlushPendingNotifications(FlushType::Layout);
 
   MOZ_TRY(UpdateSelectionAndShrinkPrintObject(aPO.get(), documentIsTopLevel));
@@ -1980,18 +2000,10 @@
 
   if (mPrintPreviewCallback) {
     const bool hasSelection = !mDisallowSelectionPrint && mSelectionRoot;
-    // Determine if there is a specified page size, and if we should set the
-    // paper orientation to match it.
-    const Maybe<bool> maybeLandscape =
-        mPrintObject->mPresShell->StyleSet()
-            ->GetDefaultPageSizeOrientation()
-            .map([](StylePageSizeOrientation o) -> bool {
-              return o == StylePageSizeOrientation::Landscape;
-            });
     mPrintPreviewCallback(PrintPreviewResultInfo(
         GetPrintPreviewNumSheets(), GetRawNumPages(), GetIsEmpty(),
         hasSelection, hasSelection && mPrintObject->HasSelection(),
-        maybeLandscape));
+        mMaybeCSSPageLandscape));
     mPrintPreviewCallback = nullptr;
   }
 